src/react/JSONTree/JSONIterableNode.js

```diff
@@ -0,0 +1,141 @@
+import React from 'react';
+import reactMixin from 'react-mixin';
+import { ExpandedStateHandlerMixin } from './mixins';
+import JSONArrow from './JSONArrow';
+import grabNode from './grab-node';
+
+const styles = {
+  base: {
+    position: 'relative',
+    paddingTop: 3,
+    paddingBottom: 3,
+    paddingRight: 0,
+    marginLeft: 14
+  },
+  label: {
+    margin: 0,
+    padding: 0,
+    display: 'inline-block'
+  },
+  span: {
+    cursor: 'default'
+  },
+  spanType: {
+    marginLeft: 5,
+    marginRight: 5
+  }
+};
+
+@reactMixin.decorate(ExpandedStateHandlerMixin)
+export default class JSONIterableNode extends React.Component {
+  defaultProps = {
+    data: [],
+    initialExpanded: false
+  };
+
+  // flag to see if we still need to render our child nodes
+  needsChildNodes = true;
+
+  // cache store for our child nodes
+  renderedChildren = [];
+
+  // cache store for the number of items string we display
+  itemString = false;
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      expanded: this.props.initialExpanded,
+      createdChildNodes: false
+    };
+  }
+
+  // Returns the child nodes for each element in the array. If we have
+  // generated them previously, we return from cache, otherwise we create
+  // them.
+  getChildNodes() {
+    if (this.state.expanded && this.needsChildNodes) {
+      let childNodes = [];
+      for (const entry of this.props.data) {
+        let key = null;
+        let value = null;
+        if (Array.isArray(entry)) {
+          [key, value] = entry;
+        } else {
+          key = childNodes.length;
+          value = entry;
+        }
+
+        let prevData;
+        if (typeof this.props.previousData !== 'undefined') {
+          prevData = this.props.previousData[key];
+        }
+        const node = grabNode(key, value, prevData, this.props.theme);
+        if (node !== false) {
+          childNodes.push(node);
+        }
+      }
+      this.needsChildNodes = false;
+      this.renderedChildren = childNodes;
+    }
+    return this.renderedChildren;
+  }
+
+  // Returns the "n Items" string for this node, generating and
+  // caching it if it hasn't been created yet.
+  getItemString() {
+    if (!this.itemString) {
+      const { data } = this.props;
+      let count = 0;
+      if (typeof data.count === 'function') {
+        count = data.count();
+      } else {
+        count = data.length;
+      }
+      this.itemString = count + ' entr' + (count !== 1 ? 'ies' : 'y');
```

Comment:
Alright, I've change it. It uses the `size` property as it's the most common (only mori doesn't have it), otherwise it counts entries through the iterator. I am not using `length` because it's deprecated in Immutable anyway and others doesn't have it at all.